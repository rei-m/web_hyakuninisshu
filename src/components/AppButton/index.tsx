import * as React from 'react';
import withStyles from '@material-ui/core/styles/withStyles';
import Icon from '@material-ui/core/Icon';
import Button from '@material-ui/core/Button';
import { appTheme } from '@src/styles/theme';

export interface Props {
  label: string;
  type: 'primary' | 'normal';
  icon?: string;
  style?: React.CSSProperties;
  onClick: () => void;
}

const styles = {
  button: {
    padding: '4px 15px',
    fontSize: '1.6rem',
    height: '40px',
  },
  normalButton: {
    color: '#182026',
    backgroundColor: '#f5f8fa',
    '&:active': {
      backgroundColor: appTheme.colorAccentActive,
    },
    '&:hover': {
      backgroundColor: '#ebf1f5',
    },
  },
  primaryButton: {
    color: '#ffffff',
    backgroundColor: appTheme.colorAccent,
    '&:active': {
      backgroundColor: appTheme.colorAccentActive,
    },
    '&:hover': {
      backgroundColor: appTheme.colorAccentHover,
    },
  },
  normalIcon: {
    marginRight: appTheme.spacing1x,
    color: '#5c7080',
  },
  primaryIcon: {
    marginRight: appTheme.spacing1x,
  },
};

const stylesProvider = () => styles;

type RenderProps = Props & {
  classes: {
    button: string;
    normalButton: string;
    primaryButton: string;
    normalIcon: string;
    primaryIcon: string;
  };
};

const AppButton = withStyles(stylesProvider)(({ label, type, icon, style, onClick, classes }: RenderProps) => {
  let buttonClassName = '';
  let iconClassName = '';
  if (type === 'primary') {
    buttonClassName = `${classes.button} ${classes.primaryButton}`;
    iconClassName = classes.primaryIcon;
  } else {
    buttonClassName = `${classes.button} ${classes.normalButton}`;
    iconClassName = classes.normalIcon;
  }

  return (
    <Button variant="outlined" onClick={onClick} className={buttonClassName} style={style}>
      {icon && (
        <Icon className={iconClassName} fontSize="small">
          {icon}
        </Icon>
      )}
      {label}
    </Button>
  );
});

export default AppButton;
