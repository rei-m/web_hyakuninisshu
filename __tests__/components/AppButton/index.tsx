import * as ReactTestRenderer from 'react-test-renderer';
import * as React from 'react';
import { shallow } from 'enzyme';
import AppButton, { Props } from '@src/components/atoms/Button';

describe('<AppButton />', () => {
  let baseProps: Props;

  beforeEach(() => {
    baseProps = {
      label: 'Test',
      type: 'primary',
    };
  });

  it('should render component when primary', () => {
    const renderer = ReactTestRenderer.create(<AppButton {...baseProps} />);
    expect(renderer.toJSON()).toMatchSnapshot();
  });

  it('should render component when normal', () => {
    const props: Props = { ...baseProps, type: 'normal' };
    const renderer = ReactTestRenderer.create(<AppButton {...props} />);
    expect(renderer.toJSON()).toMatchSnapshot();
  });

  it('should fire onClick when clicked', () => {
    const props: Props = { ...baseProps, onClick: jest.fn() };
    shallow(<AppButton {...props} />).simulate('click');
    expect(props.onClick).toHaveBeenCalled();
  });
});
