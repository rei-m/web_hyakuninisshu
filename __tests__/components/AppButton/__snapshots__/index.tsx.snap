// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AppButton /> should render component when normal 1`] = `
<button
  className="MuiButtonBase-root-32 MuiButton-root-6 MuiButton-outlined-14 Component-button-1 Component-normalButton-2"
  disabled={false}
  onBlur={[Function]}
  onContextMenu={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex="0"
  type="button"
>
  <span
    className="MuiButton-label-7"
  >
    Test
  </span>
  <span
    className="MuiTouchRipple-root-35"
  />
</button>
`;

exports[`<AppButton /> should render component when primary 1`] = `
<button
  className="MuiButtonBase-root-32 MuiButton-root-6 MuiButton-outlined-14 Component-button-1 Component-primaryButton-3"
  disabled={false}
  onBlur={[Function]}
  onContextMenu={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex="0"
  type="button"
>
  <span
    className="MuiButton-label-7"
  >
    Test
  </span>
  <span
    className="MuiTouchRipple-root-35"
  />
</button>
`;
